### PR TITLE
[WIP] Investigate potential memory bug

### LINKED ISSUE #890: Inefficient memory consumption by MeanMetric
https://github.com/PyTorchLightning/metrics/blob/3af729508289d3babf0e166d9e8405cb2b0758a2/torchmetrics/aggregation.py#L333-L408

Old average meters eats nothing compare to 16+ gb RAB by MeanMetric
```py
class AverageMeter(object):
    """Computes and stores the average and current value"""

    def __init__(self):
        self.reset()

    def reset(self):
        self.val = .0
        self.avg = .0
        self.sum = .0
        self.count = .0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count
```

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #890 
Investigating potential memory bug. Currently added test case.

## Before submitting

- [ ] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
fixed the missing package in tests :)

File: requirements/test.txt [hunk 1]
```diff
@@ -10,8 +10,10 @@ mypy>=0.790
 phmdoctest>=1.1.1
 pre-commit>=1.0
 
+psutil
 requests
 fire
 
 cloudpickle>=1.3
 scikit-learn>=0.24
+psutil
```
Comment (SkafteNicki) [inline requirements/test.txt:19]:
```suggestion
```

File: torchmetrics/metric.py [hunk 2]
```diff
@@ -247,6 +248,7 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
         cache = {attr: getattr(self, attr) for attr in self._defaults}
 
         # call reset, update, compute, on single batch
+        self._enable_grad = True  # allow grads for batch computation
```
Comment (justusschock) [inline torchmetrics/metric.py:251]: shouldn't we check whether this is enabled by `torch.is_grad_enabled`?

Comment (SkafteNicki) [reply torchmetrics/metric.py:251]: This flag controls if we allow gradient to be passed through not torch or the user.
The only time we allow this is when the user computes the metric value on the current batch.

Every time we accumulate we want to disable this such that memory does not increase over time. 
That is actual problem this PR is solving, as we currently was not disabling it if user called `metric.update`.

File: tests/bases/test_metric.py [hunk 3]
```diff
@@ -362,3 +364,42 @@ def device(self):
...
+
+
+@pytest.mark.parametrize("device", ["cpu", "cuda"])
+@pytest.mark.parametrize("requires_grad", [True, False])
+def test_constant_memory(device, requires_grad):
+    """Checks that when updating a metric the memory does not increase."""
+    if not torch.cuda.is_available() and device == "cuda":
+        pytest.skip("Test requires GPU support")
+
+    def get_memory_usage():
```
Comment (Borda) [inline tests/bases/test_metric.py:376]: Curious if this is safe in case you run tests in parallel?

Comment (SkafteNicki) [reply tests/bases/test_metric.py:376]: I think it is safe because we explicit ask for the memory consumption of the current process

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by rohitgr7 — APPROVED** [review]
